### PR TITLE
http: allow local replies to traverse the filter chain after 1xx headers

### PR DESCRIPTION
**Pull request by paul-r-gall**
Commit Message: http: allow local replies to traverse the filter chain after 1xx headers
Additional Description: If 1xx informational headers have been fully sent to the codec, and primary response headers have not yet arrived, it is safe to send a local reply through the filter chain.
Risk Level: Low
Testing: Integration tests
Docs Changes: None
Release Notes: None

### CODE REVIEW

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -987,8 +987,10 @@ void DownstreamFilterManager::sendLocalReply(
   }
 
   if (!filter_manager_callbacks_.responseHeaders().has_value() &&
-      !filter_manager_callbacks_.informationalHeaders().has_value()) {
-    // If the response has not started at all, send the response through the filter chain.
+      (!filter_manager_callbacks_.informationalHeaders().has_value() ||
+       state_.informational_response_headers_encoded_)) {
```
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:991]: also drive by, data plane changes merit (true by default) runtime guard
/wait-any

File: source/common/http/filter_manager.cc [hunk 2]
```diff
@@ -1197,6 +1199,7 @@ void FilterManager::encode1xxHeaders(ActiveStreamEncoderFilter* filter,
     }
   }
 
+  state_.informational_response_headers_encoded_ = true;
```
Comment (antoniovleonti) [inline source/common/http/filter_manager.cc:1202]: Is it possible for the filter manager to reenter a state where this should be set back to false? E.g. maybe a first 1xx response got encoded, but a second 1xx response is not yet fully processed?

**Review by antoniovleonti — COMMENTED** [review]
lgtm modulo my one comment

File: source/common/http/filter_manager.cc (hunk 1 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc]: Done.

File: source/common/http/filter_manager.cc (hunk 2 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc]: Good catch, fixed.

**Comment by antoniovleonti** [comment]
LGTM!

File: test/integration/multiplexed_integration_test.cc [hunk 3]
```diff
@@ -1667,6 +1667,51 @@ TEST_P(MultiplexedIntegrationTest, TestEncode1xxHeaders) {
...
+  auto response = std::move(encoder_decoder.second);
+
+  // Wait for the request headers to be received upstream.
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+
+  upstream_request_->encode1xxHeaders(Http::TestResponseHeaderMapImpl{{":status", "100"}});
+  response->waitFor1xxHeaders();
+  codec_client_->sendData(*request_encoder_, 10, false);
+  ASSERT_TRUE(response->waitForEndStream());
```
[the remark this answers is not in the document]
Comment (paul-r-gall) [reply test/integration/multiplexed_integration_test.cc:1706]: is that not present on line 1709? `EXPECT_EQ(response->headers().getStatusValue(), "500");`

File: source/common/http/filter_manager.cc [hunk 4]
```diff
@@ -987,8 +987,12 @@ void DownstreamFilterManager::sendLocalReply(
   }
 
   if (!filter_manager_callbacks_.responseHeaders().has_value() &&
-      !filter_manager_callbacks_.informationalHeaders().has_value()) {
-    // If the response has not started at all, send the response through the filter chain.
+      (!filter_manager_callbacks_.informationalHeaders().has_value() ||
+       (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.local_reply_traverses_filter_chain_after_1xx") &&
+        !(state_.filter_call_state_ & FilterCallState::IsEncodingMask)))) {
```
[the remark this answers is not in the document]
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc:993]: The test at https://github.com/envoyproxy/envoy/blob/64036b42e1581df185d02858d0e5d48b8aaaef97/test/integration/multiplexed_integration_test.cc#L1641 checks that 1xx local reply skips the filter chain.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**